include/render.hpp

```diff
@@ -17,148 +19,165 @@
+  for (int i = val; i > 0; --i) {
+	if ((val % i) == 0)
+	  return i;
   }
+  return 0;
 }
 
 // Render function to call the render kernel
 template <int width, int height, int samples>
-void render(sycl::queue queue, color* fb_data, const hittable_t* hittables,
-            int num_hittables, camera& cam) {
-  constexpr auto num_pixels = width * height;
-  auto const depth = 50;
-  auto frame_buf = sycl::buffer<color, 1>(fb_data, sycl::range<1>(num_pixels));
-  auto hittables_buf =
-      sycl::buffer<hittable_t, 1>(hittables, sycl::range<1>(num_hittables));
-  // Submit command group on device
-  queue.submit([&](sycl::handler& cgh) {
-    // Get memory access
-    auto frame_ptr = frame_buf.get_access<sycl::access::mode::write>(cgh);
-    auto hittables_ptr =
-        hittables_buf.get_access<sycl::access::mode::read>(cgh);
-    // Construct kernel functor
+void render(sycl::queue& queue, std::array<color, width * height>& fb,
+			std::vector<hittable_t>& hittables, camera& cam) {
+  auto constexpr depth = 50;
+  const auto nb_hittable = hittables.size();
+  auto frame_buf =
+	  sycl::buffer<color, 2>(fb.data(), sycl::range<2>(height, width));
+  auto hittables_buf = sycl::buffer<hittable_t, 1>(hittables.data(),
+												   sycl::range<1>(nb_hittable));
 
-    auto render_kernel = pixel_renderer<width, height, samples, depth>(
-        frame_ptr, hittables_ptr, num_hittables, cam);
-    executor<width, height>(cgh, render_kernel);
+  const auto& dev = queue.get_device();
+  const auto comp_unit =
+	  (buildparams::use_single_task)
+		  ? 1
+		  : dev.get_info<sycl::info::device::max_compute_units>();
+  const auto wg_size =
+	  (buildparams::use_single_task)
+		  ? 1
+		  : dev.get_info<sycl::info::device::max_work_group_size>();
+
+  const auto parallel_exec = comp_unit * wg_size;
+
+  const auto total_work = width * height;
+
+  const auto ideal_pg_height = height / parallel_exec;
+  const auto pg_height = (ideal_pg_height == 0) ? 1 : ideal_pg_height;
+  const auto pg_width =
+	  (ideal_pg_height == 0)
+		  ? find_immediate_least_divider<width>(total_work / parallel_exec)
+		  : width;
+  const auto nb_pg_line = width / pg_width;
+  const auto required_pg_line =
+	  height / ideal_pg_height + ((height % ideal_pg_height) ? 1 : 0);
+  const auto required_pg = required_pg_line * nb_pg_line;
+
+  // Submit command group on device
+  queue.submit([=, &hittables_buf, &frame_buf, &cam](sycl::handler& cgh) {
+	auto fb_acc = frame_buf.get_access<sycl::access::mode::discard_write>(cgh);
+	auto hittables_acc =
+		hittables_buf.get_access<sycl::access::mode::read>(cgh);
+	hittable_t const* hittable_ptr = hittables_acc.get_pointer();
+	color* fb_ptr = fb_acc.get_pointer();
+	const auto global = sycl::range<1>(required_pg);
+	const auto local = sycl::range<1>(wg_size);
+	cgh.parallel_for_work_group(global, local, [=](sycl::group<1> g) {
+	  g.parallel_for_work_item([&](sycl::h_item<1> item) {
+		LocalPseudoRNG rng;
```

Comment:
You need probably to initialize the seed with something depending on the id. Perhaps the concatenation of `item.get_global_id()[0]` and `item.get_global_id()[1]`?

Reply:
The iteration list is now 1D so that would not work well I guess :-)
I did a concatenation of global id and local id. 
How ever, due to the grouping of multiple pixels handling in one work item, not each pixel start the computation with the same generator state, even if initialized with the default state, and results were not "obviously lacking of randomness".